### ISSUE TITLE
Ipopt in R under Linux - a bug or complilation error?

### ISSUE THREAD

**Issue by svigerske** [coin-or/Ipopt#91]
Issue created by migration from Trac.

Original creator: anton

Original creation time: 2009-02-19 11:40:27

Assignee: ipopt-team

Version: 3.5

I have written an interface to Ipopt, which enables its use in the R environment (http://cran.r-project.org/). This helps me in solving nonlinear regression problems. This R-ipopt interface is written in C++, and it is built under Linux as a shared object libRipopt.so, which depends on libraries libipopt.so and libR.so, and is explicitly loaded by the R executable. I have built it also under Windows, with a similar design. The bundle R-Ripopt-ipopt worked fine with Windows, but under Linux it sometimes works, and sometimes it crashes with a “caught segfault” message. The Valgrind Memcheck tool has revealed that the crash happens inside the ipopt part of the code. Here is an exapmle of the Valgrind report:

```
==11082== Invalid write of size 4
==11082==    at 0xAD47D36: Ipopt::Journal::SetPrintLevel(Ipopt::EJournalCategory, Ipopt::EJournalLevel) (in /home/anton/CoinIpopt/build/lib/libipopt.so.0.0.0)
==11082==    by 0xAC0A1E2: Ipopt::IpoptApplication::OpenOutputFile(std::string, Ipopt::EJournalLevel) (in /home/anton/CoinIpopt/build/lib/libipopt.so.0.0.0)
==11082==    by 0xAC11BFC: Ipopt::IpoptApplication::Initialize(std::istream&) (in /home/anton/CoinIpopt/build/lib/libipopt.so.0.0.0)
==11082==    by 0xAC1621D: Ipopt::IpoptApplication::Initialize(std::string) (in /home/anton/CoinIpopt/build/lib/libipopt.so.0.0.0)
==11082==    by 0xA9A4DE8: do_Ripopt (Ripopt.cpp:71)
==11082==    by 0x4ED6316: (within /usr/lib/R/lib/libR.so)
==11082==    by 0x4EFE2AD: Rf_eval (in /usr/lib/R/lib/libR.so)
==11082==    by 0x4EFF67F: (within /usr/lib/R/lib/libR.so)
==11082==    by 0x4FDCF86: (within /usr/lib/R/lib/libR.so)
==11082==    by 0x4EFE063: Rf_eval (in /usr/lib/R/lib/libR.so)
==11082==    by 0x4F00277: (within /usr/lib/R/lib/libR.so)
==11082==    by 0x4EFE063: Rf_eval (in /usr/lib/R/lib/libR.so)
==11082==  Address 0x18 is not stack'd, malloc'd or (recently) free'd
```

Could you please help to resolve this problem? I should add that this happens under Debian GNU/Linux x86_64. Maybe, the problem is actually hidden inside R (and hardly in Ripopt, as I first thought), but I really don’t know. I can send you a copy of my source code together with the g++ options I used, if required. Unfortunately, I cannot send the statistical data that I processed.

I hope that this report will be useful for the further development of ipopt, which I personally find a great product.

Best regards,
Anton Tyutin

(changed by @andrea5w at 2009-02-19 15:49:46)

**Comment by svigerske** [comment]
Comment by @andrea5w created at 2009-02-19 15:49:46

Hi Anton,

First of all, I'm very happy to hear that you are using Ipopt with R and are developing an interface to R.  I have actually never used R, but it is quite popular among some colleagues.

As for the segmentation fault that you are reporting, I don't see from the valgrind output what the problem could be.  We have done many valgrind runs with Ipopt in different settings, and this didn't show up.  In some circumstances (like in matlab or the modeling environment GAMS) one has to be careful with the output, since the standard output or writing into files is somehow overloaded in those settings and one has to create extra "Journalists" that use program dependent (e.g., matlab specific) output functions.  Maybe this is also the case with what you see?

Well, I noticed in the stack that you report that the error happens in a method OpenOutputFile which seems to be only called if you are asking Ipopt to write an output file (using the output_file option).  Do you still get the error if you don't ask Ipopt to generate this output file?

Also, do you want to post the few lines of do_Ripopt that setup the IpoptApplication object and initialize it?

Andreas

**Comment by svigerske** [comment]
Comment by @andrea5w created at 2009-02-19 15:49:46

Changing assignee from ipopt-team to @andrea5w.

**Comment by svigerske** [comment]
Comment by @andrea5w created at 2009-02-19 15:49:46

Changing status from new to assigned.

**Comment by svigerske** [comment]
Comment by anton created at 2009-02-23 11:29:45

Hi Andreas,

Thank you for your feedback. I have found the cause of the crash. It was an incorrect name for the output file. (As a newcomer to Linux I missed the point that the ‘~’ sign as a substitute for the home directory may not work outside the shell.) Ipopt failed to open the file, created a null pointer file_jrnl in method IpoptApplication::OpenOutputFile, and then crashed when trying to write to array file_jrnl->print_levels_ in method Journal::SetPrintLevel. 

I propose to insert a check of the validity of the file_jrnl pointer before calling SetPrintLevel – in the same way as it is already done for pointer stdout_jrnl in method IpoptApplication::Initialize. Perhaps, it would be even more consistent to include such a check for every pointer to a Journal object created by Journalist::AddFileJournal method. In this case ipopt would not crash but merely generate empty output. This helped me to find the error, and it might help others too.

Also thank you for the hint with creating new journalists – I will try this. I had a problem with standard output in R under Windows indeed, and solved that by redefining methods FileJournal::PrintImpl, but I would prefer a more elegant solution. 

Best regards, 
Anton

**Comment by svigerske** [comment]
Comment by @andrea5w created at 2009-02-23 15:57:07

Thanks, Anton, for catching this.  I just added a bugfix that will return an error if an invalid output file name was given.  Now the IpoptApplication::Initialize method should return Invalid_Option.  The fix is so far only in trunk, and will be promoted later to stable and a release.  If you urgently need this in a release, let me know and I can make one.

By the way, someone contacted me regarding an Ipopt interface to R; I told him about this ticket and suggested to get in touch with you.  It would of course be great if you could make your R interface generally available; do you have plans to do that?  If so, we should advertise this on the Ipopt projects page.

Thanks again for the bug report, and I would be happy to hear about progress of your R interface!

Andreas

**Comment by svigerske** [comment]
Comment by anton created at 2009-02-26 15:51:09

Hi Andreas,

Both I (the developer) and the company where I work (the owner) are willing to make the Ripopt code publicly available. After some revision of the code, I will place it somewhere on the web.

There are two versions of Ripopt, one is for Linux, and the other one is for Windows. I plan to publish the version for Linux first, because it is simpler. The current version of Ripopt for Windows involves some tricks related to file and standard output. To fix the output, I changed the code of Journalist.cpp. I believe that any changes in the core Ipopt code are better avoided by any Ipopt extension aimed at public usage. I would like to eventually release a version of Ripopt which fits both Linux and Windows and is free of "dirty tricks".

Now let me announce the main Ripopt features. 
1) Ripopt is designed to replicate the functionality of the C++ interface to Ipopt. Some advanced features of the C++ interface are not supported by Ripopt, but this may be fixed in the future.
2) Ripopt does not extend the functionality of Ipopt; in particular, it does not offer automatic computation of derivatives – the user is responsible for providing R functions for computing not only the values of objective and constraint functions, but also the values of their first and (possibly) second derivatives. 
3) The main part of Ripopt is written in C++ and is build as a DLL; though this DLL can be used directly by the R user, an R wrapper is supplied for the user’s convenience. This additional R code ensures correct type conversion between R and C++, and performs a simple test of the problem description.
4) Data are sent between R and Ipopt by reference; unnecessary copying is avoided.

As already mentioned, I use Ripopt primarily to solve regression problems. For this purpose, I have developed an extension to Ripopt, which solves the following box-constrained nonlinear least squares problem:
Sum{y,,i,,-f(x,,i,,,p) : i=1,…,n}^2^ -> min,,p,,,   subject to p,,L,,(x,y) <= p <= p,,U,,(x,y),
where n is the number of observations, y,,i,,’s are the observations of the regressand (real numbers), x,,i,,’s are observations of the regressor (may be of an arbitrary type), p is a vector of regression parameters (real numbers), and f is a given real-valued function, which must be twice continuously differentiable with respect to p. This tool is called RipoptRegression (a better name is welcome), and will be published too.

Best, Anton
